---
title: "`r if (exists('reportTitle')) reportTitle else 'SUSHI Report'`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: true
    includes:
     in_header: !expr system.file("templates/fgcz_header.html", package="ezRun", lib.loc=.libPaths())
    css: !expr system.file("templates/fgcz.css", package="ezRun", lib.loc=.libPaths())
editor_options: 
  chunk_output_type: inline
---

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## PCAMDS_Result {.tabset}

```{r setup, include=FALSE}
# input for this report: sce
knitr::opts_chunk$set(echo = TRUE)
library(plotly)
library(magrittr)
library(adegenet)
library(ade4)
library(gdsfmt)
library(SNPRelate)
library(ggplot2)
library(shiny)
library(shinythemes)
library(shinydashboard)
library(Rtsne)
# knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, knitr.table.format = "html")

### PCA setup
pca <- readRDS("PCA.rds")
grouping_vars <- readRDS("grouping_vars.rds")
n_pcs <- pca$nf # number of (> 0) principal components

eig_sum <- sum(pca$eig)
pca_varprop <- pca$eig/eig_sum 
pca_varprop <- pca_varprop[1:n_pcs]

pca_tab <- data.frame(grouping_vars, pca$li, stringsAsFactors = FALSE, row.names = NULL)

tab_varprop <- as.data.frame(t(pca_varprop), stringsAsFactors = FALSE)
PC_indeces <- seq(1+ncol(grouping_vars), ncol(pca_tab))

for (i in 1:n_pcs){
  colnames(pca_tab)[i+ncol(grouping_vars)] <- paste0("PC", i)
  colnames(tab_varprop)[i] <- paste0("PC", i)
}

### MDS setup
mds <- read.csv("plink.mds", sep="")

n_dim <- ncol(mds)-ncol(grouping_vars)-1   # number of dimensions kept in mds
mds_tab <- data.frame(grouping_vars, mds[, (ncol(grouping_vars)+2):ncol(mds)], stringsAsFactors = FALSE, row.names = NULL)

### t-SNE setup
dist_matrix <- read_tsv("plink.dist", col_names = F)

max_perplexity <- floor((nrow(dist_matrix)/3) - 1)
tsne <- Rtsne(dist_matrix, perplexity=max_perplexity, check_duplicates = FALSE, is_distance = TRUE)

tsne_tab <- data.frame(grouping_vars, tsne$Y)

```

### PCA
```{r PCA, fig.width=14, fig.height=12, echo=FALSE, message=FALSE, warning=FALSE}
plot_pca <- ggplot(data = pca_tab, aes(x = PC1, y = PC2, color = Population, text = pca_tab[,1])) +
  geom_point() +
  # geom_point_interactive(aes(x = wt, y = qsec, color = disp,
  #   tooltip = sample.id, data_id = sample.id)) +
  # geom_text(hjust=0, vjust=0, label=tab$sample.id, label.size=0.1) +
  theme_classic() +
  xlab(paste0("PC1 (", format(round(pca_varprop[1]*100, 1), nsmall = 1), "%)" )) +
  ylab(paste0("PC2 (", format(round(pca_varprop[2]*100, 1), nsmall = 1), "%)" )) +
  labs(color = "Population")


# this works, but messes up the grouping
ggplotly(plot_pca, tooltip = "text")
```


### MDS
```{r MDS, fig.width=14, fig.height=12, echo=FALSE, message=FALSE, warning=FALSE}
plot_mds <- ggplot(data = mds_tab, aes(x = C1, y = C2, color = Population, text = mds_tab[,1])) +
  geom_point() +
  # geom_point_interactive(aes(x = wt, y = qsec, color = disp,
  #   tooltip = sample.id, data_id = sample.id)) +
  # geom_text(hjust=0, vjust=0, label=mds_tab$Sample, label.size=0.1) +
  theme_classic() +
  xlab("Coordinate 1") +
  ylab("Coordinate 2") +
  labs(color = "Population")

ggplotly(plot_mds, tooltip = "text")
```


### t-SNE
```{r tsne, fig.width=14, fig.height=12, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE}
### load data for testing
vcf <- read.vcfR("~/sushi_project_JB/data/test_vcf_dataset/ragi_highcov_sa0001_1k.vcf.gz")
genind <- vcfR2genind(vcf)
grouping_vars <- read.delim("~/sushi_project_JB/data/test_vcf_dataset/populations.txt")
# pop(genind) <- populations_txt$Population
genofile <- snpgdsOpen("~/git/ezRun/output_data/snp.gds")

snp_dist <- snpgdsDiss(genofile, sample.id=NULL, snp.id=NULL, autosome.only=FALSE,
    remove.monosnp=TRUE, maf=NaN, missing.rate=NaN, num.thread=1, verbose=TRUE)


tsne <- Rtsne(snp_dist$diss, perplexity=10, check_duplicates = FALSE) # You can change the value of perplexity and see how the plot changes

## Generate the t_SNE plot
# par(mfrow=c(1,2))
# plot(tsne$Y, col = "blue", pch = 19, cex = 1)
# # plot(tsne$Y, col = "black", bg= meta_data$louvain, pch = 21, cex = 1)
# plot(tsne$Y, col = "black", bg=grouping_vars$Population , pch = 21, cex = 1)


tsne_tab <- data.frame(grouping_vars, tsne$Y)


plot_tsne <- ggplot(data = tsne_tab, aes(x = X1, y = X2, color = Population, text = tsne_tab[,1])) +
  geom_point() +
  # geom_point_interactive(aes(x = wt, y = qsec, color = disp,
  #   tooltip = sample.id, data_id = sample.id)) +
  # geom_text(hjust=0, vjust=0, label=mds_tab$Sample, label.size=0.1) +
  theme_classic() +
  xlab("Coordinate 1") +
  ylab("Coordinate 2") +
  labs(color = "Population")
  
# plot_tsne

ggplotly(plot_tsne, tooltip = "text")
```

### t-SNE with plink distance-matrix
```{r tsne_plink, fig.width=14, fig.height=12, echo=FALSE, message=FALSE, warning=FALSE}
plot_tsne <- ggplot(data = tsne_tab, aes(x = X1, y = X2, color = Population, text = tsne_tab[,1])) +
  geom_point() +
  # geom_point_interactive(aes(x = wt, y = qsec, color = disp,
  #   tooltip = sample.id, data_id = sample.id)) +
  # geom_text(hjust=0, vjust=0, label=mds_tab$Sample, label.size=0.1) +
  theme_classic() +
  xlab("Coordinate 1") +
  ylab("Coordinate 2") +
  labs(color = "Population")

ggplotly(plot_tsne, tooltip = "text")
```


### t-SNE function; try perplexity
```{r tsne_function, fig.width=14, fig.height=12, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# plink_dist2 <- read_tsv("~/sushi_project_JB/dist_test2.dist")
plink_dist3 <- read_tsv("~/sushi_project_JB/dist_test3.dist", col_names = F)

tsne_fun <- function(dist, perplexity) {
  tsne <- Rtsne(plink_dist3, perplexity=perplexity, check_duplicates = FALSE, is_distance = TRUE) # You can change the value of perplexity and see how the plot changes

  tsne_tab <- data.frame(grouping_vars, tsne$Y)
  
  
  plot_tsne <- ggplot(data = tsne_tab, aes(x = X1, y = X2, color = Population, text = tsne_tab[,1])) +
    geom_point() +
    # geom_point_interactive(aes(x = wt, y = qsec, color = disp,
    #   tooltip = sample.id, data_id = sample.id)) +
    # geom_text(hjust=0, vjust=0, label=mds_tab$Sample, label.size=0.1) +
    theme_classic() +
    xlab("Coordinate 1") +
    ylab("Coordinate 2") +
    labs(color = "Population")
    
  # plot_tsne
  
  # p <- ggplotly(plot_tsne, tooltip = "text")
  return(plot_tsne)
  
}


p_i <- tsne_fun(plink_dist3, 10)
ggplotly(p_i, tooltip = "text")


```


<!-- ### Input Dataset -->
<!-- ```{r, echo=FALSE, message=FALSE} -->
<!-- # ezInteractiveTableRmd(values=ans4Report[["dataset"]]) -->
<!-- ``` -->

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
